### DISCUSSION TITLE
Using an FPGA to decode EEG data

### DISCUSSION THREAD

**Discussion by danielflood** [Xilinx/finn#1456] [Show and tell]
Hey guys, 

I just finished working on a project that uses FINN to compile a Deep CNN model with 250k+ params. It performs well even down at 2-bit precision and I got it fit it on my Arty Z720 where it can process over 100fps. The prediction task is to decode EEG data and predict which hand the subject is clenching.

To add FINN as a dependency I created my own dockerfile to wrap the FINN base image and install my own requirements. I then created my own version of the run-docker.sh file so that it opens in my own notebooks directory and follows a similar flow to the provided examples.

You can checkout my repo [here](https://github.com/danielflood/fpga-eeg).

Ps, here's a screenshot of some interesting results on the affects of quantization on prediction accuracy.
<img width="790" height="590" alt="output" src="https://github.com/user-attachments/assets/50d73214-65d1-457a-bb5b-f0fee1fea1a7" />

**Comment by fpjentzsch** [comment]
Very cool! Thanks for sharing this!

**Comment by FELLOUH-Ayoub** [comment]
thank you for sharing that with us. it's very interesting 
i have some questions if you allow.
1- i've tried to do same for a costumized MLP but i noticed when i let the last FC layer without a RELU activation after i got synthesis error. (i mean how to deal with the last layer which is supposed to pass to softmax)
2- for a costum dataset, how to run inference in PYNQ (should i write a validate.py or to just modify the generated one)
3-if we can discuss more in discord, i'll be so grateful
thank you

**Reply by danielflood** [reply]
Hey,
1) With FINN I usually just end the network with the final FC layer. That gives you the logits for each class. You can then apply softmax on the software side, or simply take the arg-max since softmax doesn’t change the predicted class.
2) For a custom dataset, the easiest path is to start by running driver.py on a single sample and make sure the end-to-end flow works. I added a notebook in my repo that shows this process:
https://github.com/danielflood/fpga-eeg/blob/main/notebooks/pynq-deployment/3_validate-model-with-pynq.ipynb
3) Sure, how do I contact you? Or feel free to ask more questions here :)
